packages/fx-core/src/component/code/tab/deploy.ts

```diff
@@ -0,0 +1,145 @@
+
+    let changed = false;
+    await forEachFileAndDir(
+      componentPath,
+      (itemPath, stats) => {
+        const relativePath = path.relative(componentPath, itemPath);
+        if (relativePath && referenceTime < stats.mtime && (!filter || filter(relativePath))) {
+          changed = true;
+          return true;
+        }
+      },
+      folderFilter
+    );
+
+    return changed;
+  }
+
+  private static async getDeploymentInfo(
+    componentPath: string,
+    envName: string,
+    telemetryReporter?: TelemetryReporter
+  ): Promise<DeploymentInfo | undefined> {
+    const deploymentDir = path.join(componentPath, PathConstants.deploymentInfoFolder);
+    const deploymentInfoPath = path.join(deploymentDir, PathConstants.deploymentInfoFile);
+
+    try {
+      const deploymentInfoJson = await fs.readJSON(deploymentInfoPath);
+      if (!deploymentInfoJson) {
+        return undefined;
+      }
+      return deploymentInfoJson[envName];
+    } catch {
+      telemetryReporter?.sendTelemetryEvent(TelemetryEvent.DeploymentInfoNotFound);
+      return undefined;
+    }
+  }
+
+  private static async getLastBuildTime(
+    componentPath: string,
+    envName: string,
+    telemetryReporter?: TelemetryReporter
+  ): Promise<Date | undefined> {
+    const deploymentInfoJson = await FrontendDeployment.getDeploymentInfo(
+      componentPath,
+      envName,
+      telemetryReporter
+    );
+    return deploymentInfoJson?.lastBuildTime
+      ? new Date(deploymentInfoJson.lastBuildTime)
+      : undefined;
+  }
+
+  private static async getLastDeploymentTime(
+    componentPath: string,
+    envName: string
+  ): Promise<Date | undefined> {
+    const deploymentInfoJson = await FrontendDeployment.getDeploymentInfo(componentPath, envName);
+    return deploymentInfoJson?.lastDeployTime
+      ? new Date(deploymentInfoJson.lastDeployTime)
+      : undefined;
+  }
+
+  public static async saveDeploymentInfo(
+    componentPath: string,
+    envName: string,
+    deploymentInfo: DeploymentInfo
+  ): Promise<void> {
+    const deploymentDir = path.join(componentPath, PathConstants.deploymentInfoFolder);
+    const deploymentInfoPath = path.join(deploymentDir, PathConstants.deploymentInfoFile);
+
+    await fs.ensureDir(deploymentDir);
+    let deploymentInfoJson: any = {};
+    try {
+      deploymentInfoJson = await fs.readJSON(deploymentInfoPath);
+    } catch {
+      // Failed to read info file, which doesn't block deployment
+    }
+
+    deploymentInfoJson[envName] ??= {};
+    deploymentInfoJson[envName].lastBuildTime =
+      deploymentInfo.lastBuildTime ?? deploymentInfoJson[envName].lastBuildTime;
+    deploymentInfoJson[envName].lastDeployTime =
+      deploymentInfo.lastDeployTime ?? deploymentInfoJson[envName].lastDeployTime;
+
+    try {
+      await fs.writeJSON(deploymentInfoPath, deploymentInfoJson);
+    } catch {
+      // Failed to write deployment info, which doesn't block deployment
```

Comment:
Does some telemetry need to be added here to monitor?

Reply:
I am not sure whether we need. But you remind me of that we need skip-deploy telemetry event which I ignored. I will add that at next PR.